ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,269 @@
+        return first(result)  # type: ignore
+
+    def __setitem__(self, key: TKey, value: TValue) -> None:
+        # updates LRU cache
+        try:
+            self.__getitem__(key)
+        except KeyError:
+            self._insert_item(key, value)
+        else:
+            self._update_item(key, value)
+
+    def __getitem__(self, key: TKey) -> TValue:
+        # updates LRU cache
+        serialized_key = self.key_encoder(key)
+        lookup_result = self._fetch_single_query(
+            "SELECT value FROM cache WHERE key=?;", (serialized_key,),
+        )
+
+        if not lookup_result:
+            raise KeyError(str(key))
+
+        (value,) = lookup_result
+
+        deserialized_value = self.value_decoder(value)
+
+        # update cache
+        self.__delitem__(key)
+        self._insert_item(key, deserialized_value)
+
+        return deserialized_value
+
+    def __delitem__(self, key: TKey) -> None:
+        serialized_key = self.key_encoder(key)
+        result = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE key=?;", (serialized_key,),
+        )
+
+        if not result:
+            raise KeyError(str(key))
+
+        node = Node(*result)
+
+        # delete key from cache
+        self._execute(
+            "DELETE FROM cache WHERE key=?;", (node.key,),
+        )
+
+        # update any nrefs/prefs in cache
+        nref_node = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE nref=?;", (node.key,),
+        )
+        pref_node = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE pref=?;", (node.key,),
+        )
+
+        if nref_node and pref_node:
+            nref = Node(*nref_node)
+            pref = Node(*pref_node)
+            self._execute(
+                "UPDATE cache SET nref=? WHERE key=?;", (pref.key, nref.key),
+            )
+            self._execute(
+                "UPDATE cache SET pref=? WHERE key=?;", (nref.key, pref.key),
+            )
+        elif nref_node:
+            nref = Node(*nref_node)
+            self._execute(
+                "UPDATE cache SET nref=? WHERE key=?;", (None, nref.key),
+            )
+        elif pref_node:
+            pref = Node(*pref_node)
+            self._execute(
+                "UPDATE cache SET pref=? WHERE key=?;", (None, pref.key),
+            )
+
+    def _insert_item(self, key: TKey, value: TValue) -> None:
+        # insert new item into map and cache
+        serialized_key = self.key_encoder(key)
+        serialized_value = self.value_encoder(value)
+        if self.is_empty:
+            new_pref = None
+            new_nref = None
+
+            self._execute(
+                "INSERT INTO cache VALUES (?, ?, ?, ?);",
+                (serialized_key, serialized_value, new_pref, new_nref),
+            )
+
+        else:
+            # evict least recently used key/value
+            if self.is_full:
```

Comment:
If you changed this to `while self.is_full:` then it seems like it would account for the situation where someone re-uses the same database with a different max size.